src/Workspaces/CSharpTest/EmbeddedLanguages/RegularExpressions/CSharpRegexParserTests.cs

```diff
@@ -0,0 +1,337 @@
+            {
+                TryParseTree(
+                    stringText.Substring(0, start) +
+                    stringText.Substring(start + 1, stringText.Length - (start + 1)),
+                    options, conversionFailureOk: true,
+                    allowIndexOutOfRange,
+                    allowNullReference,
+                    allowOutOfMemory);
+            }
+        }
+
+        private (SyntaxToken, RegexTree, ImmutableArray<VirtualChar>) JustParseTree(
+            string stringText, RegexOptions options, bool conversionFailureOk)
+        {
+            var token = GetStringToken(stringText);
+            var allChars = _service.TryConvertToVirtualChars(token);
+            if (allChars.IsDefault)
+            {
+                Assert.True(conversionFailureOk, "Failed to convert text to token.");
+                return (token, null, allChars);
+            }
+
+            var tree = RegexParser.TryParse(allChars, options);
+            if (tree != null)
+            {
+
+            }
+
+            return (token, tree, allChars);
+        }
+
+        private RegexTree TryParseTree(
+            string stringText, RegexOptions options,
+            bool conversionFailureOk,
+            bool allowIndexOutOfRange,
+            bool allowNullReference,
+            bool allowOutOfMemeory)
+        {
+            var (token, tree, allChars) = JustParseTree(stringText, options, conversionFailureOk);
+            if (tree == null)
+            {
+                Assert.True(allChars.IsDefault);
+                return null;
+            }
+
+            CheckInvariants(tree, allChars);
+
+            Regex regex = null;
+            try
+            {
+                regex = new Regex(token.ValueText, options);
+            }
+            catch (IndexOutOfRangeException) when (allowIndexOutOfRange)
+            {
+                // bug with .net regex parser.  Can happen with patterns like: (?<-0
+                Assert.NotEmpty(tree.Diagnostics);
+                return tree;
+            }
+            catch (NullReferenceException) when (allowNullReference)
+            {
+                // bug with .net regex parser.  can happen with patterns like: (?(?S))
+                return tree;
+            }
+            catch (OutOfMemoryException) when (allowOutOfMemeory)
+            {
+                // bug with .net regex parser.  can happen with patterns like: a{2147483647,}
+                return tree;
+            }
+            catch (ArgumentException ex)
+            {
+                Assert.NotEmpty(tree.Diagnostics);
+
+                // Ensure the diagnostic we emit is the same as the .Net one.  Note: we can only
+                // do this in en-US as that's the only culture where we control the text exactly
+                // and can ensure it exactly matches Regex.  We depend on localization to do a 
+                // good enough job here for other languages.
```

Comment:
That makes me think: since we want the same messages as Regex in all languages, can we populate the translated strings from those in corefx? Or can we direct the i12n folks on how to do that?

Reply:
I'll file a bug on doing that.  It seems like it would have to be someone from roslyn communicating with i12n on the best way to accomplish that.